### PR TITLE
fix: Implement GetEngineLogs in Kubernete backend

### LINKED ISSUE #998: Kurtosis engine logs - kubernetes - unimplemented panic
### What's your CLI version?

0.80.17

### Description & steps to reproduce

`kurtosis engine logs` outputs:
```shell
INFO[2023-07-26T22:56:02+02:00] Engine running in Kubernetes cluster, to connect to the engine from outside the cluster run 'kurtosis gateway' to open a local gateway to the engine
panic: implement me

goroutine 1 [running]:
github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend.(*KubernetesKurtosisBackend).GetEngineLogs(0x140006bb9b8?, {0x103455464?, 0x140005b8270?}, {0x140003ee000?, 0x20?})
	/root/project/container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/kubernetes_kurtosis_backend.go:41 +0x2c
github.com/kurtosis-tech/kurtosis/container-engine-lib/lib/backend_impls/metrics_reporting.(*MetricsReportingKurtosisBackend).GetEngineLogs(0x1035c2803?, {0x103ea7a60?, 0x140005c8420?}, {0x1400028c3c0, 0x20})
	/root/project/container-engine-lib/lib/backend_impls/metrics_reporting/metrics_reporting_kurtosis_backend.go:89 +0x38
github.com/kurtosis-tech/kurtosis/cli/cli/commands/engine/logs.run({0x103ea7a60, 0x140005c8420}, {0x103ebd2e8, 0x140005d7f90}, {0xe?, 0x10231bc60?}, {0x103ea7a60?, 0x140005c83f0?}, 0x103ab2580?, 0x103e86d78?)
	/root/project/cli/cli/commands/engine/logs/logs.go:65 +0x14c
github.com/kurtosis-tech/kurtosis/cli/cli/command_framework/highlevel/engine_consuming_kurtosis_command.(*EngineConsumingKurtosisCommand).getRunFunc.func1({0x103ea7a60, 0x140005c8420}, 0x0?, 0x104e55cc0?)
	/root/project/cli/cli/command_framework/highlevel/engine_consuming_kurtosis_command/engine_consuming_kurtosis_command.go:233 +0x198
github.com/kurtosis-tech/kurtosis/cli/cli/command_framework/lowlevel.(*LowlevelKurtosisCommand).MustGetCobraCommand.func2(0x14000359b00?, {0x104e55cc0, 0x0, 0x0})
	/root/project/cli/cli/command_framework/lowlevel/lowlevel_kurtosis_command.go:294 +0x330
github.com/spf13/cobra.(*Command).execute(0x14000359b00, {0x104e55cc0, 0x0, 0x0})
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:940 +0x5c8
github.com/spf13/cobra.(*Command).ExecuteC(0x104e05260)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:1068 +0x35c
github.com/spf13/cobra.(*Command).Execute(...)
	/go/pkg/mod/github.com/spf13/cobra@v1.7.0/command.go:992
main.main()
	/root/project/cli/cli/main.go:52 +0x80
```

Note: the `kurtosis gateway` command is running in another terminal. And I have one enclave running on kubernetes cluster.

### Desired behavior

Show the actual engine logs.

### What is the severity of this bug?

Painful; this is causing significant friction in my workflow.

### PR DESCRIPTION
**Pull request by leoporoli**
## Description:
Implement GetEnginLogs in Kubernetes backend, this feature was missing for Kubernetes

## Is this change user-facing?
YES

## References (if applicable):
Fix #998 

```
kurtosis git:lporoli/ticket-998 ❯ cli/cli/scripts/launch-cli.sh engine logs /tmp/leo/enginelogs                                                                                                       ✹
WARN[2023-07-27T14:29:13-03:00] An error occurred getting the running engine's version; you may be running an out-of-date engine version
INFO[2023-07-27T14:29:13-03:00] Engine running in Kubernetes cluster, to connect to the engine from outside the cluster run 'kurtosis gateway' to open a local gateway to the engine
INFO[2023-07-27T14:29:13-03:00] Dumped engine logs and information to directory '/tmp/leo/enginelogs'
```

```
~ ❯ cd /tmp/leo/enginelogs
enginelogs ❯ ls
kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1
enginelogs ❯ cd kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1
kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1 ❯ ls
kurtosis-engine-container.log spec.json
kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1 ❯ cat kurtosis-engine-container.log
2023-07-27T17:10:08.324950170Z DEBU[2023-07-27T17:10:08Z][enclave_pool.go:CreateEnclavePool] The enclave pool won't be activated due the pool size value is equal to zero
2023-07-27T17:10:08.325362720Z INFO[2023-07-27T17:10:08Z][main.go:runMain] Running server...
2023-07-27T17:10:08.326622241Z DEBU[2023-07-27T17:10:08Z][enclave_pool.go:destroyIdleEnclavesFromPreviousRuns] Destroying idle enclaves created before '2023-07-27 17:10:08.321544426 +0000 UTC m=+0.212799251'...
2023-07-27T17:25:14.780664962Z DEBU[2023-07-27T17:25:14Z][server.go:func1] Received gRPC request to method '/engine_api.EngineService/GetEngineInfo' with args:
2023-07-27T17:25:14.784280251Z DEBU[2023-07-27T17:25:14Z][server.go:func1] gRPC request to method '/engine_api.EngineService/GetEngineInfo' succeeded with response:
2023-07-27T17:25:14.784321368Z engine_version:"6816d1-dirty"
2023-07-27T17:28:50.062793076Z DEBU[2023-07-27T17:28:50Z][server.go:func1] Received gRPC request to method '/engine_api.EngineService/GetEngineInfo' with args:
2023-07-27T17:28:50.062834191Z DEBU[2023-07-27T17:28:50Z][server.go:func1] gRPC request to method '/engine_api.EngineService/GetEngineInfo' succeeded with response:
2023-07-27T17:28:50.062837794Z engine_version:"6816d1-dirty"
2023-07-27T17:29:00.169823525Z DEBU[2023-07-27T17:29:00Z][server.go:func1] Received gRPC request to method '/engine_api.EngineService/GetEngineInfo' with args:
2023-07-27T17:29:00.169860561Z DEBU[2023-07-27T17:29:00Z][server.go:func1] gRPC request to method '/engine_api.EngineService/GetEngineInfo' succeeded with response:
2023-07-27T17:29:00.169864226Z engine_version:"6816d1-dirty"
2023-07-27T17:29:00.207184113Z DEBU[2023-07-27T17:29:00Z][server.go:func1] Received gRPC request to method '/engine_api.EngineService/Clean' with args:
2023-07-27T17:29:00.207218459Z should_clean_all:true
2023-07-27T17:29:00.422581880Z INFO[2023-07-27T17:29:00Z][enclave_manager.go:Clean] Successfully removed the enclaves
2023-07-27T17:29:00.422733055Z INFO[2023-07-27T17:29:00Z][enclave_manager.go:Clean] Enclave Uuid 'b9db5d0bb57043ca91ee38d74b330dcc'
2023-07-27T17:29:00.423865005Z DEBU[2023-07-27T17:29:00Z][server.go:func1] gRPC request to method '/engine_api.EngineService/Clean' succeeded with response:
2023-07-27T17:29:00.423905830Z removed_enclave_name_and_uuids:{name:"divine-chasm"  uuid:"b9db5d0bb57043ca91ee38d74b330dcc"}
2023-07-27T17:29:13.397697646Z DEBU[2023-07-27T17:29:13Z][server.go:func1] Received gRPC request to method '/engine_api.EngineService/GetEngineInfo' with args:
2023-07-27T17:29:13.397735767Z DEBU[2023-07-27T17:29:13Z][server.go:func1] gRPC request to method '/engine_api.EngineService/GetEngineInfo' succeeded with response:
2023-07-27T17:29:13.397739514Z engine_version:"6816d1-dirty"
kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1 ❯ cat spec.json
{
  "volumes": [
    {
      "name": "kube-api-access-69xvt",
      "projected": {
        "sources": [
          {
            "serviceAccountToken": {
              "expirationSeconds": 3607,
              "path": "token"
            }
          },
          {
            "configMap": {
              "name": "kube-root-ca.crt",
              "items": [
                {
                  "key": "ca.crt",
                  "path": "ca.crt"
                }
              ]
            }
          },
          {
            "downwardAPI": {
              "items": [
                {
                  "path": "namespace",
                  "fieldRef": {
                    "apiVersion": "v1",
                    "fieldPath": "metadata.namespace"
                  }
                }
              ]
            }
          }
        ],
        "defaultMode": 420
      }
    }
  ],
  "containers": [
    {
      "name": "kurtosis-engine-container",
      "image": "kurtosistech/engine:6816d1-dirty",
      "ports": [
        {
          "name": "grpc",
          "containerPort": 9710,
          "protocol": "TCP"
        }
      ],
      "env": [
        {
          "name": "SERIALIZED_ARGS",
          "value": "{\"grpcListenPortNum\":9710,\"logLevelStr\":\"debug\",\"imageVersionTag\":\"6816d1-dirty\",\"metricsUserId\":\"552f582c3481a4cd0d1def434f7231c204ae4fe04598bffe0d932292ff8f409a\",\"didUserAcceptSendingMetrics\":false,\"kurtosisBackendType\":\"kubernetes\",\"kurtosisBackendConfig\":{},\"onBastionHost\":false,\"poolSize\":0}"
        }
      ],
      "resources": {},
      "volumeMounts": [
        {
          "name": "kube-api-access-69xvt",
          "readOnly": true,
          "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount"
        }
      ],
      "terminationMessagePath": "/dev/termination-log",
      "terminationMessagePolicy": "File",
      "imagePullPolicy": "IfNotPresent"
    }
  ],
  "restartPolicy": "Never",
  "terminationGracePeriodSeconds": 30,
  "dnsPolicy": "ClusterFirst",
  "serviceAccountName": "kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1",
  "serviceAccount": "kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1",
  "nodeName": "k3d-k3s-default-server-0",
  "securityContext": {},
  "imagePullSecrets": [
    {
      "name": "kurtosis-image"
    }
  ],
  "schedulerName": "default-scheduler",
  "tolerations": [
    {
      "key": "node.kubernetes.io/not-ready",
      "operator": "Exists",
      "effect": "NoExecute",
      "tolerationSeconds": 300
    },
    {
      "key": "node.kubernetes.io/unreachable",
      "operator": "Exists",
      "effect": "NoExecute",
      "tolerationSeconds": 300
    }
  ],
  "priority": 0,
  "enableServiceLinks": true,
  "preemptionPolicy": "PreemptLowerPriority"
}%
kurtosis-engine--d868d9bac0e44e81a0f8a4c5da1a9ca1 ❯
```

### CODE REVIEW

**Comment by leoporoli** [comment]
@h4ck3rk3y don't forget this guy, :),  thanks!

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/engine_functions/engine_logs.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+	if len(namespaces) > 1 {
+		return stacktrace.NewError("Found more than one engine's namespace; this is a bug in Kurtosis")
+	}
+
+	for engineGuidStr, namespacesForId := range namespaces {
+		engineGuid := engine.EngineGUID(engineGuidStr)
+		if len(namespacesForId) > 1 {
+			return stacktrace.NewError(
+				"Expected at most one namespace to match engine GUID '%v', but got '%v'",
+				len(namespacesForId),
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/engine_functions/engine_logs.go]: the arguments should be the other way around

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/engine_functions/engine_logs.go [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+				engineGuidStr,
+			)
+		}
+
+		engineNamespace := namespacesForId[0]
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/engine_functions/engine_logs.go:54]: are we sure that there's at least one name space? can the name spaces be a nil slice

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/engine_functions/engine_logs.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+
+		if len(podsForId) == 0 {
+			return stacktrace.NewError(
+				"Expected to find one engine pod in namespace '%v' for engine with GUID '%v' "+
+					"but none was found",
+				namespaceName,
+				engineGuid,
+			)
+		}
+		if len(podsForId) > 1 {
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/engine_functions/engine_logs.go:88]: for later but we should probably create a generic somewhere that throws an error if the thing isn't size 1; we do this is 0, 1 or more than one check at a few places

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/shared_helpers/shared_helpers.go [hunk 4]
```diff
@@ -670,6 +701,95 @@ func GetMatchingUserServiceObjectsAndKubernetesResourcesByServiceName(
...
+	if len(resultErrorsByPodName) > 0 {
+		combinedErr := BuildCombinedError(resultErrorsByPodName, dumpPodErrorTitle)
+		return combinedErr
+	}
+	return nil
+}
+
+// This is a helper function that will take multiple errors, each identified by an ID, and format them together
+// If no errors are returned, this function returns nil
+func BuildCombinedError(errorsById map[string]error, titleStr string) error {
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/shared_helpers/shared_helpers.go:763]: we should probably take this out of k8s and put it in some place generic later on

File: container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/shared_helpers/shared_helpers.go [hunk 5]
```diff
@@ -682,3 +802,106 @@ func getEnclaveMatchLabels() map[string]string {
...
+			namespaceName,
+			podName,
+			containerName,
+			shouldFollowPodLogsWhenDumping,
+			shouldAddTimestampsWhenDumpingPodLogs,
+		)
+		if err != nil {
+			return stacktrace.Propagate(
+				err,
+				"An error occurred dumping logs of container '%v' in pod '%v' in namespace '%v'",
```
Comment (h4ck3rk3y) [inline container-engine-lib/lib/backend_impls/kubernetes/kubernetes_kurtosis_backend/shared_helpers/shared_helpers.go]: maybe we can be more specific and say an error occurred getting logs; as this isn't dumping them just yet

**Review by h4ck3rk3y — APPROVED** [review]

**Pull request merged**